### PR TITLE
RYA-377 Rya Streams

### PR DESCRIPTION
**Pull request by kchilton2**
This is a first draft of the Rya Streams feature. It's a single node Kafka Streams implementation of an incremental SPARQL processor. It supports most of the query use cases of the Fluo app we've written, but it does not have any periodic support yet. We still need to hook historic import and PCJ/Construct output into the source Rya instance using Kafka Connect.

### CODE REVIEW

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-rya-master-with-optionals-pull-requests/580/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-rya-master-with-optionals-pull-requests/590/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-rya-master-with-optionals-pull-requests/591/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-rya-master-with-optionals-pull-requests/594/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-rya-master-with-optionals-pull-requests/595/

File: common/pom.xml [hunk 1]
```diff
@@ -34,6 +34,8 @@ under the License.
 
     <modules>
         <module>rya.api</module>
+        <module>rya.api.model</module>
+        <module>rya.api.function</module>
```
Comment (meiercaleb) [inline common/pom.xml]: Based on the contents of this module, it seems like rya.api.evaluation would a better name.  Naming the module function suggests that the module consists entirely of Filter Functions.  Just a nit tho.

File: common/rya.api.function/src/main/java/org/apache/rya/api/function/aggregation/AverageFunction.java [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+ * Update the {@link AggregationState}'s average if the child Binding Set contains the binding name
+ * that is being averaged by the {@link AggregationElement}.
+ */
+@DefaultAnnotation(NonNull.class)
+public final class AverageFunction implements AggregationFunction {
+    private static final Logger log = LoggerFactory.getLogger(AverageFunction.class);
+
+    @Override
+    public void update(final AggregationElement aggregation, final AggregationState state, final VisibilityBindingSet childBindingSet) {
+        checkArgument(aggregation.getAggregationType() == AggregationType.AVERAGE, "The AverageFunction only accepts AVERAGE AggregationElements.");
```
Comment (meiercaleb) [inline common/rya.api.function/src/main/java/org/apache/rya/api/function/aggregation/AverageFunction.java]: Preconditions.

File: common/rya.api.function/src/main/java/org/apache/rya/api/function/aggregation/CountFunction.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+/**
+ * Increments the {@link AggregationState}'s count if the child Binding Set contains the binding name
+ * that is being counted by the {@link AggregationElement}.
+ */
+@DefaultAnnotation(NonNull.class)
+public final class CountFunction implements AggregationFunction {
+    @Override
+    public void update(final AggregationElement aggregation, final AggregationState state, final VisibilityBindingSet childBindingSet) {
+        checkArgument(aggregation.getAggregationType() == AggregationType.COUNT, "The CountFunction only accepts COUNT AggregationElements.");
```
Comment (meiercaleb) [inline common/rya.api.function/src/main/java/org/apache/rya/api/function/aggregation/CountFunction.java]: Preconditions

File: common/rya.api.function/src/main/java/org/apache/rya/api/function/aggregation/MaxFunction.java [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+ * maxed by the {@link AggregationElement}.
+ */
+@DefaultAnnotation(NonNull.class)
+public final class MaxFunction implements AggregationFunction {
+
+    private final ValueComparator compare = new ValueComparator();
+
+    @Override
+    public void update(final AggregationElement aggregation, final AggregationState state, final VisibilityBindingSet childBindingSet) {
+        checkArgument(aggregation.getAggregationType() == AggregationType.MAX, "The MaxFunction only accepts MAX AggregationElements.");
```
Comment (meiercaleb) [inline common/rya.api.function/src/main/java/org/apache/rya/api/function/aggregation/MaxFunction.java]: Preconditions

File: common/rya.api.function/src/main/java/org/apache/rya/api/function/aggregation/MinFunction.java [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+ * mined by the {@link AggregationElement}.
+ */
+@DefaultAnnotation(NonNull.class)
+public final class MinFunction implements AggregationFunction {
+
+    private final ValueComparator compare = new ValueComparator();
+
+    @Override
+    public void update(final AggregationElement aggregation, final AggregationState state, final VisibilityBindingSet childBindingSet) {
+        checkArgument(aggregation.getAggregationType() == AggregationType.MIN, "The MinFunction only accepts MIN AggregationElements.");
```
Comment (meiercaleb) [inline common/rya.api.function/src/main/java/org/apache/rya/api/function/aggregation/MinFunction.java]: Preconditions

File: extras/rya.streams/api/pom.xml [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+        <dependency>
+            <groupId>org.slf4j</groupId>
+            <artifactId>slf4j-api</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>com.google.guava</groupId>
+            <artifactId>guava</artifactId>
+        </dependency>
+        
+        <dependency>
```
Comment (meiercaleb) [inline extras/rya.streams/api/pom.xml]: Included twice.

File: extras/rya.streams/kafka/src/main/java/org/apache/rya/streams/kafka/processors/ProcessorResult.java [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+
+import edu.umd.cs.findbugs.annotations.DefaultAnnotation;
+import edu.umd.cs.findbugs.annotations.NonNull;
+
+/**
+ * Represents a value that is emitted from a Rya Streams {@link Processor}. We can't just emit a
+ * {@link VisibilityBindingSet} because some downstream processors require more information about
+ * which upstream processor is emitting the result in order to do their work.
+ * </p>
+ * Currently there are only two types processors:
```
Comment (meiercaleb) [inline extras/rya.streams/kafka/src/main/java/org/apache/rya/streams/kafka/processors/ProcessorResult.java]: types of

File: extras/rya.streams/kafka/src/main/java/org/apache/rya/streams/kafka/processors/join/CloseableIterator.java [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+import edu.umd.cs.findbugs.annotations.DefaultAnnotation;
+import edu.umd.cs.findbugs.annotations.NonNull;
+
+/**
+ * An {@link Iterator} that is also {@link AutoCloseable}.
+ *
+ * @param <T> - The type of elements that will be iterated over.
+ */
+@DefaultAnnotation(NonNull.class)
+public interface CloseableIterator<T> extends Iterator<T>, AutoCloseable { }
```
Comment (meiercaleb) [inline extras/rya.streams/kafka/src/main/java/org/apache/rya/streams/kafka/processors/join/CloseableIterator.java]: Do we really need another CloseableIterator interface?

File: extras/rya.streams/kafka/src/main/java/org/apache/rya/streams/kafka/processors/join/JoinProcessorSupplier.java [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+            final List<String> allVars,
+            final ProcessorResultFactory resultFactory) throws IllegalArgumentException {
+        super(resultFactory);
+        this.stateStoreName = requireNonNull(stateStoreName);
+        this.join = requireNonNull(join);
+        this.joinVars = requireNonNull(joinVars);
+        this.allVars = requireNonNull(allVars);
+
+        if(!allVars.subList(0, joinVars.size()).equals(joinVars)) {
+            throw new IllegalArgumentException("All vars must be lead by the join vars, but it did not. " +
```
Comment (meiercaleb) [inline extras/rya.streams/kafka/src/main/java/org/apache/rya/streams/kafka/processors/join/JoinProcessorSupplier.java]: Awkward Exception message.

File: extras/rya.streams/kafka/src/main/java/org/apache/rya/streams/kafka/processors/join/JoinStateStore.java [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+ */
+@DefaultAnnotation(NonNull.class)
+public interface JoinStateStore {
+
+    /**
+     * Store a {@link VisibilityBindingSet} based on the side it was emitted from.
+     *
+     * @param result - The result whose value will be stored. (not null)
+     */
+    public void store(BinaryResult result);
```
Comment (meiercaleb) [inline extras/rya.streams/kafka/src/main/java/org/apache/rya/streams/kafka/processors/join/JoinStateStore.java]: Yet another way of serializing VisiblityBindingsets?  Is this really necessary?

File: extras/rya.streams/api/src/main/java/org/apache/rya/streams/api/interactor/StartQuery.java [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+import org.apache.rya.streams.api.exception.RyaStreamsException;
+
+import edu.umd.cs.findbugs.annotations.DefaultAnnotation;
+import edu.umd.cs.findbugs.annotations.NonNull;
+
+/**
+ * Starts processing a query in Rya Streams.
+ */
+@DefaultAnnotation(NonNull.class)
+public interface StartQuery {
```
[the remark this answers is not in the document]
Comment (kchilton2) [reply extras/rya.streams/api/src/main/java/org/apache/rya/streams/api/interactor/StartQuery.java]: Start query indicates the query execution system that is running in the background should ensure that query is being processed. The RunQuery command will execute the query using the invoking thread.

File: extras/rya.streams/kafka/src/main/java/org/apache/rya/streams/kafka/processors/join/JoinProcessorSupplier.java (hunk 9 above)
Comment (kchilton2) [reply extras/rya.streams/kafka/src/main/java/org/apache/rya/streams/kafka/processors/join/JoinProcessorSupplier.java]: What would you rather it say?

File: extras/rya.streams/kafka/src/main/java/org/apache/rya/streams/kafka/processors/join/CloseableIterator.java (hunk 8 above)
Comment (kchilton2) [reply extras/rya.streams/kafka/src/main/java/org/apache/rya/streams/kafka/processors/join/CloseableIterator.java]: Yes.

info.aduna.iteration.CloseableIteration does not extend AutoCloseable, so I can't use try with resources blocks on it.

org.eclipse.rdf4j.common.iteration.CloseableIteration is part of rdf4j. We use an sesame right now, so I can not access that class. But even if I could, it uses the info.aduna.iteration.Iteration interface to perform iteration, which is not for each loop compatible.

This class only uses Java interfaces. It allows me to use both the try with resources and the for each loop features of Java. If you know of another class in our project that uses only the Java interfaces, I'll gladly swap over.

File: extras/rya.streams/kafka/src/main/java/org/apache/rya/streams/kafka/processors/join/JoinStateStore.java (hunk 10 above)
Comment (kchilton2) [reply extras/rya.streams/kafka/src/main/java/org/apache/rya/streams/kafka/processors/join/JoinStateStore.java]: What would you like me to use instead?

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-rya-master-with-optionals-pull-requests/615/

File: extras/rya.streams/kafka/src/main/java/org/apache/rya/streams/kafka/processors/join/CloseableIterator.java (hunk 8 above)
Comment (meiercaleb) [reply extras/rya.streams/kafka/src/main/java/org/apache/rya/streams/kafka/processors/join/CloseableIterator.java]: I thought that you implemented another instance of CloseableIterator for similar reasons when you were doing the Fluo work.  Did that get deleted?

Comment (kchilton2) [reply extras/rya.streams/kafka/src/main/java/org/apache/rya/streams/kafka/processors/join/CloseableIterator.java]: Somebody deleted that.

Comment (kchilton2) [reply extras/rya.streams/kafka/src/main/java/org/apache/rya/streams/kafka/processors/join/CloseableIterator.java]: Or at least I thought they did, let me check again.

File: extras/rya.streams/kafka/src/main/java/org/apache/rya/streams/kafka/processors/join/JoinProcessorSupplier.java (hunk 9 above)
Comment (meiercaleb) [reply extras/rya.streams/kafka/src/main/java/org/apache/rya/streams/kafka/processors/join/JoinProcessorSupplier.java]: "The vars List must start with the joinVars List, but it did not" or "The joinVars List must be a prefix of the vars List".  I understand that "start with" and "prefix" are not perfectly well defined in the context of List, but "lead by" stood out as being awkward to me.

File: extras/rya.streams/kafka/src/main/java/org/apache/rya/streams/kafka/processors/join/CloseableIterator.java (hunk 8 above)
Comment (kchilton2) [reply extras/rya.streams/kafka/src/main/java/org/apache/rya/streams/kafka/processors/join/CloseableIterator.java]: I found a static one inside of PrecomputedJoinStorage. Where do you want me to move it to? That's in the secondary indexing project, which I do not depend on. I could move it to the api project.

File: extras/rya.streams/kafka/src/main/java/org/apache/rya/streams/kafka/processors/join/JoinStateStore.java (hunk 10 above)
Comment (meiercaleb) [reply extras/rya.streams/kafka/src/main/java/org/apache/rya/streams/kafka/processors/join/JoinStateStore.java]: It's just that there are at least two implementations of Serializers for VisiblityBindingSet int eh code base.  Granted that they might be for specific serialization/encoder interfaces.  Were you at least able to reuse VisibilityBindingSetSerDe in the implementation of this interface?

File: extras/rya.streams/kafka/src/main/java/org/apache/rya/streams/kafka/processors/join/JoinProcessorSupplier.java (hunk 9 above)
Comment (kchilton2) [reply extras/rya.streams/kafka/src/main/java/org/apache/rya/streams/kafka/processors/join/JoinProcessorSupplier.java]: Updated.

File: common/rya.api.function/src/main/java/org/apache/rya/api/function/aggregation/AverageFunction.java (hunk 2 above)
Comment (kchilton2) [reply common/rya.api.function/src/main/java/org/apache/rya/api/function/aggregation/AverageFunction.java]: Done.

File: common/rya.api.function/src/main/java/org/apache/rya/api/function/aggregation/CountFunction.java (hunk 3 above)
Comment (kchilton2) [reply common/rya.api.function/src/main/java/org/apache/rya/api/function/aggregation/CountFunction.java]: Done.

File: common/rya.api.function/src/main/java/org/apache/rya/api/function/aggregation/MaxFunction.java (hunk 4 above)
Comment (kchilton2) [reply common/rya.api.function/src/main/java/org/apache/rya/api/function/aggregation/MaxFunction.java]: Done.

File: common/rya.api.function/src/main/java/org/apache/rya/api/function/aggregation/MinFunction.java (hunk 5 above)
Comment (kchilton2) [reply common/rya.api.function/src/main/java/org/apache/rya/api/function/aggregation/MinFunction.java]: Done.

File: common/rya.api.function/src/main/java/org/apache/rya/api/function/aggregation/SumFunction.java [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+ * Add to the {@link AggregationState}'s sum if the child Binding Set contains the binding name
+ * that is being summed by the {@link AggregationElement}.
+ */
+@DefaultAnnotation(NonNull.class)
+public final class SumFunction implements AggregationFunction {
+    private static final Logger log = LoggerFactory.getLogger(SumFunction.class);
+
+    @Override
+    public void update(final AggregationElement aggregation, final AggregationState state, final VisibilityBindingSet childBindingSet) {
+        checkArgument(aggregation.getAggregationType() == AggregationType.SUM, "The SumFunction only accepts SUM AggregationElements.");
```
[the remark this answers is not in the document]
Comment (kchilton2) [reply common/rya.api.function/src/main/java/org/apache/rya/api/function/aggregation/SumFunction.java]: Done.

File: extras/rya.streams/api/pom.xml (hunk 6 above)
Comment (kchilton2) [reply extras/rya.streams/api/pom.xml]: Done.

File: extras/rya.streams/kafka/src/main/java/org/apache/rya/streams/kafka/processors/join/JoinStateStore.java (hunk 10 above)
Comment (kchilton2) [reply extras/rya.streams/kafka/src/main/java/org/apache/rya/streams/kafka/processors/join/JoinStateStore.java]: I'm using whatever, org.apache.kafka.streams.state.KeyValueStore uses by default to serialized the VisibilityBindingSets to the Value section of the KeyValueStore.

If you want to make sure that code makes sense, the implementing class is KeyValueJoinStateStore.

File: extras/rya.streams/kafka/src/main/java/org/apache/rya/streams/kafka/processors/ProcessorResult.java (hunk 7 above)
Comment (kchilton2) [reply extras/rya.streams/kafka/src/main/java/org/apache/rya/streams/kafka/processors/ProcessorResult.java]: Done.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-rya-master-with-optionals-pull-requests/621/<h2>Build result: FAILURE</span></h2>[...truncated 2.32 MB...]	at hudson.model.AbstractBuild$AbstractBuildExecution.run(AbstractBuild.java:504)	at hudson.model.Run.execute(Run.java:1724)	at hudson.maven.MavenModuleSetBuild.run(MavenModuleSetBuild.java:543)	at hudson.model.ResourceController.execute(ResourceController.java:97)	at hudson.model.Executor.run(Executor.java:421)Caused by: java.io.IOException: Unexpected termination of the channel	at hudson.remoting.SynchronousCommandTransport$ReaderThread.run(SynchronousCommandTransport.java:77)Caused by: java.io.EOFException	at java.io.ObjectInputStream$PeekInputStream.readFully(ObjectInputStream.java:2675)	at java.io.ObjectInputStream$BlockDataInputStream.readShort(ObjectInputStream.java:3150)	at java.io.ObjectInputStream.readStreamHeader(ObjectInputStream.java:859)	at java.io.ObjectInputStream.<init>(ObjectInputStream.java:355)	at hudson.remoting.ObjectInputStreamEx.<init>(ObjectInputStreamEx.java:48)	at hudson.remoting.AbstractSynchronousByteArrayCommandTransport.read(AbstractSynchronousByteArrayCommandTransport.java:35)	at hudson.remoting.SynchronousCommandTransport$ReaderThread.run(SynchronousCommandTransport.java:63)ERROR: H26 is offline; cannot locate JDK 1.8 (latest)ERROR: H26 is offline; cannot locate Maven 3.3.9ERROR: H26 is offline; cannot locate JDK 1.8 (latest)ERROR: H26 is offline; cannot locate Maven 3.3.9ERROR: H26 is offline; cannot locate JDK 1.8 (latest)ERROR: H26 is offline; cannot locate Maven 3.3.9Setting status of 007eb0a76e593d0ecb2b9c6e0fc4092fc372891a to FAILURE with url https://builds.apache.org/job/incubator-rya-master-with-optionals-pull-requests/621/ and message: 'FAILURE 'Using context: Jenkins: clean package -Pgeoindexing

File: common/pom.xml (hunk 1 above)
Comment (kchilton2) [reply common/pom.xml]: Done.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-rya-master-with-optionals-pull-requests/619/

File: extras/rya.streams/kafka/src/main/java/org/apache/rya/streams/kafka/processors/join/CloseableIterator.java (hunk 8 above)
Comment (kchilton2) [reply extras/rya.streams/kafka/src/main/java/org/apache/rya/streams/kafka/processors/join/CloseableIterator.java]: Moved to api and updated all of the references.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-rya-master-with-optionals-pull-requests/622/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-rya-master-with-optionals-pull-requests/623/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-rya-master-with-optionals-pull-requests/626/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-rya-master-with-optionals-pull-requests/625/<h2>Failed Tests: <span class='status-failure'>3</span></h2><h3><a name='incubator-rya-master-with-optionals-pull-requests/org.apache.rya:rya.prospector' /><a href='https://builds.apache.org/job/incubator-rya-master-with-optionals-pull-requests/625/org.apache.rya$rya.prospector/testReport'>incubator-rya-master-with-optionals-pull-requests/org.apache.rya:rya.prospector</a>: <span class='status-failure'>3</span></h3><ul><li><a href='https://builds.apache.org/job/incubator-rya-master-with-optionals-pull-requests/625/org.apache.rya$rya.prospector/testReport/org.apache.rya.prospector.mr/ProspectorTest/testCount/'><strong>org.apache.rya.prospector.mr.ProspectorTest.testCount</strong></a></li><li><a href='https://builds.apache.org/job/incubator-rya-master-with-optionals-pull-requests/625/org.apache.rya$rya.prospector/testReport/org.apache.rya.prospector.service/ProspectorServiceEvalStatsDAOTest/testCount/'><strong>org.apache.rya.prospector.service.ProspectorServiceEvalStatsDAOTest.testCount</strong></a></li><li><a href='https://builds.apache.org/job/incubator-rya-master-with-optionals-pull-requests/625/org.apache.rya$rya.prospector/testReport/org.apache.rya.prospector.service/ProspectorServiceEvalStatsDAOTest/testNoAuthsCount/'><strong>org.apache.rya.prospector.service.ProspectorServiceEvalStatsDAOTest.testNoAuthsCount</strong></a></li></ul>

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-rya-master-with-optionals-pull-requests/639/

**Pull request closed without merging**